### PR TITLE
Implement compute metrics for model.py

### PR DESCRIPTION
**Pull request by Clydingus**
model.py has 3 new compute metric functions that compute metrics for the
latest output tensor that passed through the model.

2 are for attention and ffn layers and compute the inf_norm and kurtosis
of the outputs

the last is for softmax sum metric, constructs the tensor for attention
softmax output, this also supports being called as a function externally
so we can get rid of that unsafe global variable import to get
softmax_sum.

additionally wandb logging has been updated and cleaned up to implement
the 3 compute metric functions during eval stage.

attention_sums.ipynb has been updated to make use of the new 
softmax_sums compute metrics method.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -209,12 +209,9 @@ def forward(
...
 
-            if FLAGS['inspect_softmax_sum']:
-                global softmax_sum
-                if len(softmax_sum) >=12: # num_blocks hardcoded
-                    softmax_sum = []
-                sums = scores.detach().squeeze(0).sum(-1).cpu()
-                softmax_sum.append(sums)
+            # save intermediate softmax tensor for compute_softmax_metrics during eval state
+            if not self.training:
+                self.softmax_output = scores.detach().cpu()
```
Comment (photomz) [inline model.py:218]: Can we define `softmax_output` at init?

Comment (Clydingus) [reply model.py:218]: You're probably right. I referred to https://github.com/tcapelle/llama2.c/blob/af8597ca6ac0c74142951c44b8d6926e170d791a/model.py when implementing the compute metrics and the self.output was not initialized under init, which in hindsight is probably not that good practice. I'll make another commit to define them at init.

File: model.py [hunk 2]
```diff
@@ -411,3 +423,35 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None, return_logi
             return idx, torch.cat(logits_list, dim=1)
         
         return idx
+
+    # *yoink*: https://github.com/tcapelle/llama2.c/blob/af8597ca6ac0c74142951c44b8d6926e170d791a/model.py#L428
+    # the following metric computations require that the model be run in eval mode, then the respective layer outputs
+    # will be saved in order to compute the metric of the latest input passed through the model
+    def compute_attention_metrics(self) -> Tuple[List[float], List[float]]:
```
Comment (photomz) [inline model.py:430]: Don't we already have `inf_norm` and `kurtosis` metrics?

File: model.py [hunk 3]
```diff
@@ -411,3 +423,35 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None, return_logi
...
+        compute the softmax sum of the attention outputs
+        output shape: [batch size, layer number, attention head, seq len]
+        """
+        # each softmax_output should be of shape [batch size, attention head, seq len]
+        softmax_sums = [b.attention.softmax_output.sum(-1) for b in self.layers]
+        # stack together tensors from different layers, but make sure stack of layers occurs in dim=1
+        # such that final output shape is [batch size, layer number, attention head, seq len]
+        return torch.stack(softmax_sums, dim=1)
+
+    # TODO Implement compute_perplexity maybe, although this will not be as reliable as compute_perplexity() in
```
Comment (photomz) [inline model.py:455]: Why can't `compute_perplexity` compute ppl over a sequence in this file, yet the notebook can?

File: model.py [hunk 4]
```diff
@@ -411,3 +423,35 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None, return_logi
...
+    # the following metric computations require that the model be run in eval mode, then the respective layer outputs
+    # will be saved in order to compute the metric of the latest input passed through the model
+    def compute_attention_metrics(self) -> Tuple[List[float], List[float]]:
+        "compute the max inf norm and kurtosis of the attention outputs"
+        outputs = [b.attention.output for b in self.layers]
+        k = [kurtosis(o.flatten().half()) for o in outputs]
+        inf_norm = [o.abs().max().item() for o in outputs]
+        return inf_norm, k
+
+    def compute_ffn_metrics(self) -> Tuple[List[float], List[float]]:
```
Comment (photomz) [inline model.py:437]: Seems a bit redundant with `compute_attention_metrics`

File: model.py [hunk 5]
```diff
@@ -411,3 +423,35 @@ def generate(self, idx, max_new_tokens, temperature=1.0, top_k=None, return_logi
...
+        "compute the max inf norm and kurtosis of the ffn outputs"
+        outputs = [b.feed_forward.output for b in self.layers]
+        k = [kurtosis(o.flatten().half()) for o in outputs]
+        inf_norm = [o.abs().max().item() for o in outputs]
+        return inf_norm, k
+
+    def compute_softmax_metrics(self) -> torch.Tensor:
```
Comment (photomz) [inline model.py:444]: In the future, we should have pure functions as a `@staticmethod` and add PyTests.

File: train.py [hunk 6]
```diff
@@ -260,13 +260,33 @@ def compute_metrics():
                 logits = model(X, Y)
                 loss = raw_model.last_loss
             losses[k] = loss.item()
+        # TODO @Markus you mentioned this was computing metrics wrongly, so I commented it out, if this is
```
Comment (photomz) [inline train.py:263]: This computes the _wrong_ metrics, but the computation is correct. We should add more tests so we're comfy with the metrics.

File: train.py [hunk 7]
```diff
@@ -260,13 +260,33 @@ def compute_metrics():
...
+                        }
+        """
+        out[split] = {'loss': losses.mean()}
         deregister()
+
+    # compute metrics on last batch of validation
+    att_inf_norm, att_kurtosis = raw_model.compute_attention_metrics()
+    ffn_inf_norm, ffn_kurtosis = raw_model.compute_ffn_metrics()
+    for i, (att_norm, att_k, ffn_norm, ffn_k) in enumerate(zip(att_inf_norm, att_kurtosis, ffn_inf_norm, ffn_kurtosis)):
+        out[split].update({f"atten_inf_norm_{i}": att_norm,
```
Comment (photomz) [inline train.py:280]: This is a little messy, I'll clean it up

File: notebooks/attention_sums.ipynb
Comment (photomz) [inline notebooks/attention_sums.ipynb:1]: Pls ensure your tab/spaces config is the same. the diff is hard to read

**Review by photomz — COMMENTED** [review]
Thanks for keeping with my original pattern, but I feel my bookkeeping is too messy. It's getting hard to track which metrics are saved where, how, and if they're tested properly.

Let's keep this PR for reference and inspiration. I'll refactor first, then we can starting adding new metrics @Clyde013
